### ISSUE TITLE
Tabs Don't Work When Opening a Folder Instead of a Single File

### ISSUE THREAD

**Issue by bdkjones** [coteditor/CotEditor#1712]
## Environment

- CotEditor: 5.0.0 (674)
- System: macOS Version 15.0 (Build 24A335)
- Language: English

## Short Description

NSWindow tabs cannot be used if I open a folder (such as a website project) in CotEditor. The View > Tab Bar command is greyed out. CMD+T doesn't work. There is no option to select a file in the sidebar and "Open in New Tab". This severely limits CotEditor's usefulness for coding.

Tabs *ARE* available if I do File > New and simply have a blank document open. I can then create as many tabs as I'd like.

## Steps to Reproduce the Issue

1. Open a folder of files in CotEditor.
2. Try to open multiple files in separate tabs.

## Expected Result

1. Tabs should be available when I open a folder in CotEditor.
2. Double-clicking or option-clicking a file in the sidebar should open that file in a new tab (or switch to the existing tab that contains that file, if any.)

**Comment by bdkjones** [comment]
I'd really like to use CotEditor as a replacement for Panic's Nova, but I definitely need the ability to open multiple files from a folder in tabs within a single window so I can quickly jump between them. 

Even better: I'd like to open different files in split-view. Splitting the same document isn't as useful and putting two different documents side-by-side so I can reference one while coding in the other.

**Comment by 1024jp** [comment]
Yes, It's currently in spec. I'd like to solve this in the future, but I skipped it to implement it in 5.0.0.

**Comment by bdkjones** [comment]
Got it. Any idea when 5.0.0 might be available?

**Comment by 1024jp** [comment]
@bdkjones

> Any idea when 5.0.0 might be available?

Today! https://itunes.apple.com/app/coteditor/id1024640650?ls=1

**Comment by bdkjones** [comment]
First, congratulations!

Second, I misunderstood. I thought you meant "this feature is coming in 5.0" but you meant "I dropped this feature so I could get 5.0 shipped." In that case, any idea when this feature might land?

**Comment by 1024jp** [comment]
> In that case, any idea when this feature might land?

I have no idea yet. I will do that within the next year.
However, for that, I need to make my own tab views and tab system. It might be hard.
The current window tab system provided by Apple is very limited and doesn't work in this case.

**Comment by LeoNatan** [comment]
Please do not make a custom tab system. This is one of the reasons I ran away from other editors. It should be possible to implement using the Apple system; `NSWindow` exposes enough API to open tabs (menu item or user preference) or new windows (user preference) using `addTabbedWindow(_:ordered:)`. What did find impossible to implement?

**Comment by 1024jp** [comment]
Closed the wrong ticket. Reopen.

**Comment by MastroPino** [comment]
Hello! There is any update on this? 
With macOS Tahoe the problem persist

**Comment by 1024jp** [comment]
This is not an issue but by design at the moment.

**Comment by bdkjones** [comment]
That decision should be revisited. There is not another editor that fails to support multiple tabs for various files in a single folder. Even Vim does so.

This single decision caused me to abandon CotEditor. Tabs are a non-negotiable, must-have capability for anyone working on code.

**Comment by MastroPino** [comment]
Hi there! 
I downloaded the source code from GitHub (I’m a web developer but I don’t know Swift). With Claude’s help I managed to restore the use of tabs inside projects as well, and I noticed that tab management was intentionally excluded from opening a project (meant as a set of folders and files) and enabled only when opening individual files. My implementation is obviously (almost certainly) terrible, and for anyone who’s fluent in Swift it definitely won’t be hard to improve!

How can I help?

<img width="1324" height="759" alt="Image" src="https://github.com/user-attachments/assets/20072c72-9292-4b52-ab4f-7002956e00cf" />

**Comment by LeoNatan** [comment]
> With Claude’s help

Brave new world

**Comment by MastroPino** [comment]
> > With Claude’s help
> 
> Brave new world

From dev to dev: I didn’t mean to upset anyone, my bad

**Comment by LeoNatan** [comment]
Why upset? It's all good! Just amusing how these tools are so effective.

**Comment by nicolasdejong** [comment]
A solution that would work without needing tabs is to remember the scroll location of files when closing them. Then when clicking on a file in the tree on the left, the selected file will open at the correct scroll-location. It should do auto-save when switching files. Oh, and placing the cursor at the old location completes the feature. So for each file then the scroll-location and cursor-location should be stored.

The only differences with having tabs is then that tabs are a kind of a short-list of recently opened files that is not available in the project view.

**Comment by 1024jp** [comment]
Since files may be modified by other processes, I cannot be overly aggressive about preserving insertion points.
CotEditor is not an IDE; it is fundamentally a document-based application that opens a single file for a window.
The macOS application framework simply does not anticipate this kind of structure.

Still, I understand the sentiment. I will consider it long-term.

**Comment by mar3ld** [comment]
I'm evaluating multiple editors. The no tabs when working in project mode got me surprised — got me backing off. Can't even see in the browser which files are open, or that are unsaved. That looks a bit broken to me, and therefor I'm avoiding this editor for now.
Otherwise this looks like an interesting project. Not sure why the devs thinks different on this topic. Did I miss something?
7.0.2 for reference.

**Comment by 1024jp** [comment]
@mar3ld Please read my previous post. CotEditor is not the editor you’re looking for.

**Comment by mar3ld** [comment]
@1024jp Hi. Thanks for your response. Not the one I'm looking for? This editor actually has much of what I'm looking for, but I sadly need to look elsewhere as I really need to be able to quickly change between tabs for the open documents I'm working on at the moment.

(I use a text editor mostly for note taking. The ability to turn off auto save is fantastic. I think the app needs an indicator in the browser though, pointing out unsaved files.)
